parser/parser/cc/grammars/typedruby.ypp

```diff
@@ -1715,6 +1731,12 @@ lbrace_cmd_block_start:
                     {
                       $$ = driver.build.array(self, $1, $2, $3);
                     }
+                | tLBRACK aref_args array_premature_end
+                    {
+                      $$ = driver.build.array(self, $1, $2, $3);
+                      driver.diagnostics.emplace_back(dlevel::ERROR, dclass::UnterminatedToken, diagnostic::range(@1.begin, @1.end), "\"[\"");
+                      driver.rewind_and_reset(@2.end);
```

Comment:
I think that this use of `rewind_and_reset` is a little suspect. I think that if we're sold on this approach, the lexer state we want to be in once we've said that the recovery phase is done is `expr_beg`, because that's the state that the lexer would have transitioned to had it correctly seen  `]` here.

Reply:
Hmm, yeah, `beg` does make more sense here! I'll try switching to that. (Maybe it'll make some of the less-than-ideal parses you identified better---who knows?)

Reply:
I think `expr_end` might be the right state, actually...

https://github.com/sorbet/sorbet/blob/f5d8393c630cf05c158d55fb3dfabbd9d5a0d628/parser/parser/cc/lexer.rl#L2824-L2839

This is consistent with what the Ruby Hacking Guide has to say (https://whitequark.org/blog//2013/04/01/ruby-hacking-guide-ch-11-finite-state-lexer/). Whatcha think?